src/morx.rs

```diff
@@ -0,0 +1,1824 @@
+                    self.component_stack.push(glyph.clone());
+                    if self.component_stack.len() == 1 {
+                        start_pos = i; //mark the position in the buffer for the first glyph in a ligature group
+                    }
+                }
+
+                if entry_flags & PERFORM_ACTION != 0 {
+                    //Perform Action: use the ligActionIndex to process a ligature group
+
+                    end_pos = i; //mark the position in the buffer for the last glyph in a ligature group
+                    let mut action_index: usize = usize::from(entry.lig_action_index);
+                    let mut index_to_ligature: u16 = 0;
+                    let mut ligature: RawGlyph<()> = RawGlyph {
+                        unicodes: tiny_vec![[char; 1]],
+                        glyph_index: 0x0000,
+                        liga_component_pos: 0,
+                        glyph_origin: GlyphOrigin::Direct,
+                        small_caps: false,
+                        multi_subst_dup: false,
+                        is_vert_alt: false,
+                        fake_bold: false,
+                        fake_italic: false,
+                        extra_data: (),
+                        variation: None,
+                    };
+
+                    //loop through stack
+                    'stack: loop {
+                        let glyph_popped: u16;
+
+                        match self.component_stack.pop() {
+                            Some(val) => {
+                                glyph_popped = val.glyph_index;
+
+                                let mut unicodes = val.unicodes;
+                                unicodes.append(&mut ligature.unicodes);
+                                ligature.unicodes = unicodes;
+
+                                ligature.variation = val.variation;
+                            }
+                            None => return Err(ParseError::MissingValue),
+                        };
+
+                        let action: u32 = ligature_subtable.action_table.actions[action_index];
+                        action_index += 1;
+
+                        let mut offset = action & 0x3FFFFFFF; //take 30 bits
+
+                        if offset & 0x20000000 != 0 {
+                            offset |= 0xC0000000; //sign-extend it to 32 bits
+                        }
+                        let offset = offset as i32; //convert to signed integer
+
+                        let index_to_components = glyph_popped as i32 + offset;
+
+                        if index_to_components < 0 {
+                            return Err(ParseError::BadValue);
+                        }
+
+                        let index_to_component_table: usize =
+                            match usize::try_from(index_to_components) {
+                                Ok(index) => index,
+                                Err(_err) => return Err(ParseError::BadValue),
+                            };
```

Comment:
You can skip the `< 0` check as `usize::try_from` will return an `Err` if the value is negative.